Add support for structured noise (perlin) fuzzy skin

Arachnid opened the pull request:
# Description

Fixes #7467.

The existing fuzzy skin support does a good job of hiding printing artefacts, but uses a very naive algorithm based on randomly displacing points on the outside wall of an object. This means that there's no correlation between the displacements of vertically adjacent points, and this in turn gives the fuzzy skin a characteristic horizontal pattern that can accentuate rather than hide layer lines.

This PR adds libnoise, a library for creating structured noise, and uses it to enhance the fuzzy skin functionality with new options that have better aesthetics compared to the existing uniform random noise. Several noise algorithms are supported:

 - Perlin noise: A classic structured noise algorithm that creates spatially coherent noise.
 - Billow noise: A variation on perlin noise that has a more 'billowly' or cloud-like appearance.
 - Ridged Multifractal noise: Creates random ridgelike formations, or marble-like textures.
 - Voronoi: Divides space up into voronoi cells, and displaces each by a random amount. Creates a patchwork effect.

A number of parameters are provided to affect the output:
 - Feature size: Adjusts the size of the predominant features in the noise. In mm. Defaults to 1mm.
 - Octaves (all but Voronoi): Each noise algorithm uses several instances with decreasing feature size, and sums them. This controls how many instances are instantiated.
 - Persistence (Perlin, Billow only): Each octave contributes progressively less to the total. This controls the rate at which octaves' amplitude is reduced.

Existing parameters - depth and point distance - continue to be used in the expected manner.

Backwards-compatibility is provided for by having 'classic' as the default noise type when fuzzy skin is enabled.

I've included the library inline following the example of some other libraries used by Orca, as it's only available via download from Sourceforge via an interstitial. If there's a better way to do this, please let me know.

# Screenshots/Recordings/Graphs
![Screenshot from 2024-12-06 16-46-10](https://github.com/user-attachments/assets/64936204-9d16-452e-8418-91cdea4dc750)

![PXL_20241206_090346843 (1)](https://github.com/user-attachments/assets/2f266c21-d722-4cfd-a451-bddb50c17989)
[Bambulab Calibration Cube by prdak27](https://makerworld.com/en/models/164937#profileId-181146).
0.2mm point distance, 0.2mm thickness.

![PXL_20241206_090417016 (1)](https://github.com/user-attachments/assets/7a33e80b-3d22-402f-8c81-ed19b23e0a95)
0.2mm point distance, 0.2mm thickness.

![PXL_20241206_090448835 (1)](https://github.com/user-attachments/assets/7de3bb58-3552-46aa-a1e9-0b43cfb77f74)
![PXL_20241206_154823381](https://github.com/user-attachments/assets/d596a40b-3b0a-4a8a-aa12-b6ccb5b159bd)
[Fuzzy Love Teddy Bear by SamActivity](https://makerworld.com/en/models/173870#profileId-191088)
Classic: 0.8mm point distance, 0.3mm thickness.
Voronoi: 0.2mm point distance, 0.2mm thickness, 5mm feature size
All others: 0.2mm point distance, 0.3mm thickness, 3mm feature size, 5 octaves, 0.66 persistence

discip commented:
@Arachnid
Thank you! 👍
Great, this makes Orca even better!

I hope you don't mind: Here are some suggestions you might consider:
1. Maybe you could reorder and rename the entries like this:
   - Structured skin
   - Noise type
   - Point distance
   - Amplitude
   - Feature size
   - Noise octaves
   - Noise persistence
   - Apply structure to first layer

2. Also you might want to hide the non relevant entries if a specific noise type is chosen.

discip commented:
@Arachnid
Thank you very much! 👍
Great, this is absolutely fantastic and contributes to Orca being the go to slicer!

I hope you don't mind: Here are some suggestions you might consider:
1. Maybe you could reorder and rename the entries like this:
   - Structured skin
   - Noise type
   - Point distance
   - Amplitude
   - Feature size
   - Noise octaves
   - Noise persistence
   - Apply structure to first layer

2. Also you might want to hide the non relevant entries if a specific noise type is chosen.

p.s.
Would it be possible implement an option to use custom patterns (e.g.: svg) that would define the structure of the skin? Something like a brick pattern for example.

discip commented:
@Arachnid
Thank you very much! 👍
Great, this is absolutely fantastic and contributes to Orca being the go to slicer!

I hope you don't mind: Here are some suggestions you might consider:
1. Maybe you could reorder and rename the entries like this:
   - Structured skin
   - Noise type
   - Point distance
   - Amplitude
   - Feature size
   - Noise octaves
   - Noise persistence
   - Apply structure to first layer

2. You may also want to hide irrelevant entries when a particular noise type is selected.

p.s.
Would it be possible implement an option to use custom patterns (e.g.: svg) that would define the structure of the skin? Something like a brick pattern for example.

discip commented:
@Arachnid
Thank you very much! 👍
Great, this is absolutely fantastic and contributes to Orca being the go to slicer!

I hope you don't mind: Here are some suggestions you might consider:
1. Maybe you could reorder and rename the entries like this:
   - Structured skin
   - Noise type
   - Point distance
   - Amplitude
   - Feature size
   - Noise octaves
   - Noise persistence
   - Apply structure to first layer

2. You may also want to hide irrelevant entries when a particular noise type is selected.

p.s.
Would it be possible to implement an option to use custom patterns (e.g. SVG) that define the structure of the skin? For example, something like a brick pattern.

Arachnid commented:
> I hope you don't mind: Here are some suggestions you might consider:
> 
> 1. Maybe you could reorder and rename the entries like this:
>    
>    * Structured skin
>    * Noise type
>    * Point distance
>    * Amplitude
>    * Feature size
>    * Noise octaves
>    * Noise persistence
>    * Apply structure to first layer

I'm happy to consider it! I'd like to hear from the maintainers before renaming existing features, though - there may be concerns about confusion / backwards compatibility.



> 2\. You may also want to hide irrelevant entries when a particular noise type is selected.

Already done!



> p.s.
> Would it be possible to implement an option to use custom patterns (e.g. SVG) that define the structure of the skin? For example, something like a brick pattern.

This is tricky. The existing noise library generates 3 dimensional noise, meaning we can get the value for any point just by querying it. Mapping a 2d object like an SVG to a surface is much more complex, and I'm afraid it's beyond my basic computational geometry skills!

Arachnid commented:
Fixes #7467

discip commented:
@Arachnid
> Fixes #7467

If you want the linked issue to be closed with the merge of this PR, you need to put it in the initial post via an edit.

discip commented:
@Arachnid
> Fixes #7467

If you want the linked issue to be closed with the merge of this PR, you need to put it in the initial post via an edit. 😊

Noisyfox commented:
Is it possible to add the libnoise as an external dep that can be downloaded as a zip during build?

Arachnid commented:
> Is it possible to add the libnoise as an external dep that can be downloaded as a zip during build?

As far as I can tell, libnoise is only hosted on source forge, which has a browser interstitial for all downloads. I can try and bypass it for an automated download, but I was concerned that would be subject to breakage easily. 

I can also delete irrelevant files such as docs, if that would help.

Noisyfox commented:
> I can also delete irrelevant files such as docs, if that would help.

Absolutely. Now my browser freezes as soon as I switch to the Files Changed tab

discip commented:
@Noisyfox
Could we please at least reorder the entries at this point?
I think the current order looks like it was patched together.

This is what I suggest to change in **`Tab.cpp`**:
```C++
2348        optgroup->append_single_option_line("fuzzy_skin");
2349        optgroup->append_single_option_line("fuzzy_skin_noise_type");
2350        optgroup->append_single_option_line("fuzzy_skin_point_distance");
2351        optgroup->append_single_option_line("fuzzy_skin_thickness");
2352        optgroup->append_single_option_line("fuzzy_skin_scale");
2353        optgroup->append_single_option_line("fuzzy_skin_octaves");
2354        optgroup->append_single_option_line("fuzzy_skin_persistence");
2355        optgroup->append_single_option_line("fuzzy_skin_first_layer");

```

Optionally line **`2355`** `("fuzzy_skin_first_layer")` could be placed after **`2349`** `("fuzzy_skin_noise_type")`.

@Arachnid
While testing this very nice addition, I noticed that **`Multifractal`** and **`Billow`** are mixed up in the last image of the initial post. 😊

Arachnid commented:
I've reordered the options as suggested by @discip. I've also fixed the image - good catch.

I've also removed the docs folder to cut down on the number of files imported. Still happy to restructure it to download it at build time, if we have a good solution to the sourceforge hosting.

discip reviewed (approved):
## Great! 👍
Thank you!

SeeTurtlee commented:
I've been playing with this for about a week now and its super cool! i was wondering however if it would be possible to add the option to apply fuzzy skin to the top layer? i know it's technically possible with https://github.com/TengerTechnologies/Fuzzyficator, but having it as an actual feature would be awesome!

Arachnid commented:
> I've been playing with this for about a week now and its super cool! i was wondering however if it would be possible to add the option to apply fuzzy skin to the top layer? i know it's technically possible with https://github.com/TengerTechnologies/Fuzzyficator, but having it as an actual feature would be awesome!

Adding fuzzy skin to top surfaces is a whole different thing! I'd love to have it too, but I think it will require a separate pr reimplementing their hard work in C++.

Arachnid commented:
@SoftFever Appreciate you have a lot going on, but I'd hugely appreciate it if you're able to give this a review.

undingen commented on src/libslic3r/PerimeterGenerator.cpp:
```diff
@@ -1005,17 +1051,17 @@ static ExtrusionEntityCollection traverse_extrusions(const PerimeterGenerator& p
                         extrusion.junctions.clear();
 
-                        const auto fuzzy_current_segment = [&segment, &extrusion, &r]() {
+                        const auto fuzzy_current_segment = [&segment, extrusion, &r, slice_z]() {
```
I'm getting a compile error here: because `extrusion` should IMHO be passed by reference like before.

undingen commented:
Very impressive results  :100: 
I tried taking this PR as a base for experimenting with using displacement map/texture for the surface - one thing I noticed: this code is still adding some additional noise by randomizing the line start position via `random_value()` and length. Not sure if this is on purpose I guess it does not really make a difference for the implemented noise sources but when I added a call to "checkboard" I seem to notice some of this random segments. Not sure if this should be removed if not in classic mode?!?

Arachnid commented:
> one thing I noticed: this code is still adding some additional noise by randomizing the line start position via random_value() and length. Not sure if this is on purpose I guess it does not really make a difference for the implemented noise sources but when I added a call to "checkboard" I seem to notice some of this random segments. Not sure if this should be removed if not in classic mode?!?

Good catch. I left it in because I didn't want to break legacy fuzzy skin, and leaving it as-is reduced the size of the diff, while not making a meaningful difference to the results with random noise. I can see how it would affect something more deterministic though!

SoftFever commented:
@Arachnid 
Thank you!
This is very nice.

Can we move the libnoise to third-party library (deps) so that we don't have to build it every time?
Please let me know if you need any help.

SoftFever commented:
I love it!
![image](https://github.com/user-attachments/assets/5efc98a2-10f5-4b28-99e1-64ac32642c40)

Arachnid commented:
@SoftFever Glad you like it! I'm happy to change how it's built, but I could use some advice on how. When I was looking at the process originally it looked like libraries in deps had to be downloaded as part of the build process, and the only download source for this library is sourceforge, with an interstitial page.

SoftFever commented:
> @SoftFever Glad you like it! I'm happy to change how it's built, but I could use some advice on how. When I was looking at the process originally it looked like libraries in deps had to be downloaded as part of the build process, and the only download source for this library is sourceforge, with an interstitial page.

We can create a repo just like what we did for nanosvg and wxwidget.
I can you help you move it.
Give me 1 or 2 hours to work on it

SoftFever commented:
@Arachnid 
I have move the libnoise to deps and have made the necessary change.
I need your help to allow me push my changes into your branch. If you are fine with it, you can enable it here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Another option is that I can create another PR

Arachnid commented:
Amazing, thank you! Allow edits by maintainers is already ticked, so you should be able to push to my branch.

SoftFever commented:
> Amazing, thank you! Allow edits by maintainers is already ticked, so you should be able to push to my branch.

Hmm I still can't push

Arachnid commented:
Merged!

SoftFever commented:
There are some compiler error after merging with main branch.
I will fix them in a new PR along with some CICD issues.
I'll merge this PR first as the main code logic looks good to me.

Thank you very much!

Arachnid commented:
Amazing. Thank you so much!

henrivdr commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2632,6 +2641,57 @@ void PrintConfigDef::init_fff_params()
+    def->set_default_value(new ConfigOptionFloat(1.0));
+
+    def = this->add("fuzzy_skin_octaves", coInt);
+    def->label = L("Fuzzy Skin Noise Octaves");
```
@Arachnid
Hi. Skin has an uppercase S here, lower elsewhere.

henrivdr commented:
@Arachnid Can I recommend that the settings be moved to "Quality" , with it's own "Fuzzy skin" group and icon?
I think it deserves it's own group :+1: 
@SoftFever your thoughts?  Not sure what the icon should look like :-)

Arachnid commented:
@henrivdr I think that makes a lot of sense. Happy to open a PR if @SoftFever agrees!

igiannakas commented:
> @henrivdr I think that makes a lot of sense. Happy to open a PR if @SoftFever agrees!

Makes sense to have this as a separate group of settings now, from a usability perspective :) 

Under ironing may be a good place to create a new section as ironing is already collapsed if not used and fuzzy skin would be equally collapsed.

SoftFever commented:
That's a good suggestion to move it to its own group! 
However, I hope it remains in the Other's tab for two reasons:
1. People are getting used to finding the options here.
2. I'm not entirely sure it fits within the Quality category, as that tab should, in principle, only contain options that affect surface quality or accuracy. While fuzzy skin represents aesthetic changes.

igiannakas commented:
Makes sense :)

discip commented:
> 2\. I'm not entirely sure it fits within the Quality category, as that tab should, in principle, only contain options that affect surface quality or accuracy. While fuzzy skin represents aesthetic changes.

Absolutely agree!

igiannakas commented:
Just wanted to say a huge thank you for this - the new patterns look absolutely stunning!

![IMG_5484](https://github.com/user-attachments/assets/d7a64463-e3cf-47c5-87e5-6088e5c6f1d5)

Zogar89 commented:
> Just wanted to say a huge thank you for this - the new patterns look absolutely stunning!
> 
> ![IMG_5484](https://private-user-images.githubusercontent.com/59056762/409139171-d7a64463-e3cf-47c5-87e5-6088e5c6f1d5.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY0MTgzMjUsIm5iZiI6MTc0NjQxODAyNSwicGF0aCI6Ii81OTA1Njc2Mi80MDkxMzkxNzEtZDdhNjQ0NjMtZTNjZi00N2M1LTg3ZTUtNjA4OGU1YzZmMWQ1LmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNVQwNDA3MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMWRhMWMwNDAxYzI1OTIwOWJmNTkwNzA4MjkyMzUyMmQwYzMwZjM2NWRhMjAwMTA1MzUzZTZjOWQ5OWJkMWY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.uz7PsYj_kFC4bURvr0lIjV8U4fffk76LF_5Pd_5ePTc)

Please, Could you share the options you used to achieve that finish?
